src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
   CallGenerator::do_late_inline_helper();
 }
 
+static void delay_box_in_uncommon_trap(CallNode* call, Node* resproj) {
+  if (resproj != nullptr && call->is_CallStaticJava() &&
+      call->as_CallStaticJava()->is_boxing_method()) {
+    Unique_Node_List uncommon_trap_list;
+    bool no_use = true;
+    for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
+      Node* m = resproj->fast_out(i);
+      if (m->is_CallStaticJava() &&
+          m->as_CallStaticJava()->uncommon_trap_request() != 0) {
+        uncommon_trap_list.push(m);
+      } else {
+        no_use = false;
+        break;
+      }
+    }
+
+    if (no_use) {
+      GraphKit kit(call->jvms());
+      PhaseGVN& gvn = kit.gvn();
+      // delay box node in uncommon_trap runtime, treat box as a scalarized object
+      while (uncommon_trap_list.size() > 0) {
+        ProjNode* res = resproj->as_Proj();
+        Node* uncommon_trap_node = uncommon_trap_list.pop();
+
+        ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
+        int n_fields = klass->nof_nonstatic_fields();
+        assert(n_fields == 1, "the klass must be an auto-boxing klass");
+
+        uint first_ind = (uncommon_trap_node->req() - uncommon_trap_node->jvms()->scloff());
+        Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  (AllocateNode*)call,
```

Comment:
You can use call->isa_Allocate();  It utilizes node's ad-hoc RTTI  to do type casting.

Reply:
OK. I have done that.